### ISSUE TITLE
zmq::context_t dtor does not handle EINTR

### ISSUE THREAD

**Issue by andrust** [zeromq/cppzmq#171]
http://api.zeromq.org/master:zmq-ctx-destroy states:
"Return value

The zmq_ctx_destroy() function shall return zero if successful. Otherwise it shall return -1 and set errno to one of the values defined below.

Errors

EFAULT
The provided context was invalid.
EINTR
Termination was interrupted by a signal. It can be restarted if needed."

**Comment by sigiesec** [comment]
Could you provide an example usage, and state what behaviour you expect? Should the dtor retry in that case?
You are also welcome to send a PR which adds a test case, and implements the expected behaviour.

**Comment by egpbos** [comment]
I am mainly interested in handling EINTR, which I'm currently running into (as mentioned in #28). I will think about how to trigger this signal, but not sure whether I can manage, because for me it is currently a bug, so I don't fully understand it yet.

**Comment by andrust** [comment]
I don't remember the exact situation how i stumbled upon this issue, but i think a signal was interrupting the close function during shutdown.

https://github.com/zeromq/cppzmq/blob/c444c5d5e2ecd9fef2da2455809599bfa34347fd/zmq.hpp#L518-L528
```cpp
    inline ~context_t() ZMQ_NOTHROW { close(); }

    inline void close() ZMQ_NOTHROW
    {
        if (ptr == NULL)
            return;

        int rc = zmq_ctx_destroy(ptr);
        ZMQ_ASSERT(rc == 0);
        ptr = NULL;
    }
```

In above snippet, i think we should retry `zmq_ctx_destroy` when `errno == EINTR` as the documentation states. It is unclear to me what the `if needed` part means. Maybe someone, who's familiar with the C code could make it clear.

**Comment by egpbos** [comment]
I also thought just retrying is probably the right solution, but just to be sure, I dove into the C code, and now I'm not so sure anymore.

Looking through the libzmq code for `zmq_ctx_destroy`, we see that it just calls [`zmq_ctx_term`](https://github.com/zeromq/libzmq/blob/master/src/zmq.cpp#L149). There we can see that the only step that will be skipped in the `EINTR` case is the `zmq::shutdown_network()` call, but that does almost nothing. Most of the work is in [`zmq::ctx_t::terminate`](https://github.com/zeromq/libzmq/blob/master/src/ctx.cpp#L131). There we see that [terminate is built to handle retry calls](https://github.com/zeromq/libzmq/blob/master/src/ctx.cpp#L162). [A little further down](https://github.com/zeromq/libzmq/blob/master/src/ctx.cpp#L181) we see the only way this function (and thus `zmq_ctx_destroy`) will return `-1` and `errno` `EINTR` is when `_term_mailbox.recv` fails this way as well. This, in turn, only fails this way [if `_signaler.wait` fails this way](https://github.com/zeromq/libzmq/blob/master/src/mailbox.cpp#L81).

Now, here we arrive at the system calls, but it's slightly more tricky than expected:

- [The first situation that can cause `signaler_t::wait` to give `EINTR`](https://github.com/zeromq/libzmq/blob/master/src/signaler.cpp#L232) is when `pid != getpid ()`, which could happen when using `fork()`. In fact, my use case makes use of forking, so it doesn't seem like a corner case for me, but this may be just my biased view.
- Then, finally, we arrive at the actual (possibly interrupted) system calls: either `poll` or `select`, depending on build configuration. These give `EINTR` on any system call that they receive during their operation. This could be some random failure, but it could also be a deliberate signal, like pressing Ctrl-C to terminate the program. See e.g. [this blog post from @sustrik](http://250bpm.com/blog:12).

My guess is that, in the first (`pid != getpid ()`) case, there is no real hope for recovery and calling `zmq_ctx_destroy` again will just yield the same result, leading to an infinite loop. If possible, this should then be checked in the destructor as well and on this maybe print a descriptive error message, like what the comment says about the lost file descriptor. However, `pid` is a private member of `signaler_t`, so we would need to modify `zmq::context_t` to also have such a `pid` member that can detect forking. I'm not sure whether this is desired at this point, and maybe, in any case, it deserves a separate issue/discussion.

In the second case, the question by @sustrik is valid: should we retry or handle signals? My guess is that retrying is probably okay here.

- If the user is really trying to Ctrl-C (or `kill -9`, or otherwise assassinate) the program, just pressing it a few times will almost certainly kill the program in the end, because it won't always be waiting inside `poll` or `select`, and when it's not, the signal will just cause the program to end as usual.
- If some other program, like a child proces (SIGCHLD), is sending signals, I think retrying clean up is still the way to go. I'm not sure what would happen with a debugger (SIGTRAP). Would using breakpoints become broken?

So, in summary, I think the destructor should become something like:

```
inline void close() ZMQ_NOTHROW
{
    if (ptr == NULL)
        return;

    do {
        int rc = zmq_ctx_destroy(ptr);
        // TODO: decide on adding handling of broken forking:
        // if (pid != getpid()) { std::cerr << "broken forking" << std::endl; return; }
    } while (rc == -1 && errno == EINTR)
    ZMQ_ASSERT(rc == 0);
    ptr = NULL;
}
```

I'll make a PR for this.
